common/flatpak-installation.c

```diff
@@ -1024,76 +1052,114 @@ flatpak_installation_list_installed_refs_for_update (FlatpakInstallation *self,
-  g_auto(GStrv) refs_app = NULL;
-  g_auto(GStrv) refs_runtime = NULL;
+  GLNX_AUTO_PREFIX_ERROR (G_STRFUNC, error);
+  g_autoptr(GPtrArray) installed_refs = NULL;
+  g_autofree gboolean *installed_refs_used = NULL;
+  g_autoptr(GPtrArray) installed_refs_for_update = NULL;
+  g_autoptr(GHashTable) related_to_refs = NULL; /* (element-type utf8 utf8) */
+  g_autoptr(FlatpakTransaction) transaction = NULL;
+  g_autoptr(GError) local_error = NULL;
 
-  /* We clone the dir here to make sure we re-read the latest ostree repo config, in case
-     it has local changes */
-  dir = flatpak_dir_clone (dir_orig);
-  if (!flatpak_dir_maybe_ensure_repo (dir, cancellable, error))
+  installed_refs = flatpak_installation_list_installed_refs (self, cancellable, error);
+  if (installed_refs == NULL)
     return NULL;
 
-  remote_names = flatpak_dir_list_remotes (dir, cancellable, error);
-  if (remote_names == NULL)
+  /* Here we use a FlatpakTransaction to determine what needs updating, and
+   * abort it before actually doing the updates. This ensures we are consistent
+   * with the CLI update command.
+   */
+  transaction = flatpak_transaction_new_for_installation (self, cancellable, error);
+  if (transaction == NULL)
     return NULL;
 
-  remote_states = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, (GDestroyNotify)flatpak_remote_state_unref);
-  for (int i = 0; remote_names[i] != NULL; ++i)
+  for (guint i = 0; i < installed_refs->len; i++)
     {
-      const char *remote = remote_names[i];
-      g_autoptr(FlatpakRemoteState) state = NULL;
-      g_autoptr(GError) local_error = NULL;
+      FlatpakInstalledRef *ref_tmp = g_ptr_array_index (installed_refs, i);
+      g_autofree char *ref = flatpak_ref_format_ref (FLATPAK_REF (ref_tmp));
 
-      if (flatpak_dir_get_remote_disabled (dir, remote))
+      if (flatpak_transaction_add_update (transaction, ref, NULL, NULL, &local_error))
         continue;
 
-      state = flatpak_dir_get_remote_state_optional (dir, remote, FALSE, NULL, &local_error);
-      if (state == NULL)
+      if (g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_REMOTE_NOT_FOUND))
         {
-          g_debug ("Update: Failed to read remote %s: %s", remote, local_error->message);
-          continue;
+          g_debug ("%s: Unable to update %s: %s", G_STRFUNC, ref, local_error->message);
+          g_clear_error (&local_error);
+        }
+      else
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
         }
-      g_hash_table_insert (remote_states, (char *)remote, g_steal_pointer (&state));
     }
 
-  updates = g_ptr_array_new_with_free_func (g_object_unref);
+  related_to_refs = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
 
-  if (flatpak_dir_list_refs (dir, "app", &refs_app, cancellable, error))
+  g_signal_connect (transaction, "ready", G_CALLBACK (transaction_ready), &related_to_refs);
```

Comment:
This will conflict a bit with https://github.com/flatpak/flatpak/pull/3510 that runs some ops before emitting the ready call. I think we will need a way for this to disable the auto-install of the authenticator, or at least to emit a different ready before installing them so we can cancel it.

Reply:
Ah, should we wait for that to land then? This is relatively urgent since Endless 3.8.0 is in about 3 weeks.

Reply:
No, land this first but we need to take care of this conflict when we land that. #3510 needs rewriting anyway.